doc/manual/typing.md

```diff
@@ -0,0 +1,629 @@
+like Nickel.
+
+#### Row polymorphism
+
+In a configuration language, you will often find yourself handling records of
+various kinds. In a simple type system, you can hit the following issue:
+
+```nickel
+(let addTotal: {total: Num} -> {total: Num} -> Num
+  = fun r1 r2 => r1.total + r2.total in
+let r1 = {jan = 200, feb = 300, march = 10, total = jan + feb} in
+let r2 = {aug = 50, sept = 20, total = aug + sept} in
+let r3 = {may = 1300, june = 400, total = may + june} in
+{
+  partial1 = addTotal r1 r2,
+  partial2 = addTotal r2 r3,
+}) : {partial1: Num, partial2: Num}
+```
+
+```
+error: Type error: extra row `sept`
+  ┌─ repl-input-40:8:23
+  │
+8 │   partial2 = addTotal r2 r3,
+  │                       ^^ this expression
+  │
+  = The type of the expression was expected to be `{total: Num}`, which does not contain the field `sept`
+  = The type of the expression was inferred to be `{total: Num, sept: Num, aug: Num}`, which contains the extra field `sept`
+```
+
+The problem here is that for this code to run fine, the requirement of
+`addTotal` should be that both arguments have a field `total: Num`, but could
+very well have other fields, for all we care. Unfortunately, we don't know right
+now how to express this constraint. The current annotation is too restrictive,
+because it imposes that arguments have exactly one field `total: Num`, and
+nothing more.
+
+To express such constraints, Nickel features *row polymorphism*. The idea is
+similar to polymorphism, but instead of substituting a parameter for a single type,
+we can substitute a parameter for a whole sequence of field declarations, also
+referred to as rows:
+
+```nickel
+(let addTotal: forall a b. {total: Num | a} -> {total: Num | b} -> Num
+  = fun r1 r2 => r1.total + r2.total in
+let r1 = {jan = 200, feb = 300, march = 10, total = jan + feb} in
+let r2 = {aug = 50, sept = 20, total = aug + sept} in
+let r3 = {may = 1300, june = 400, total = may + june} in
+{
+  partial1 = addTotal r1 r2,
+  partial2 = addTotal r2 r3,
+}) : {partial1: Num, partial2: Num}
+```
+
+Result:
+```
+{partial1 = 570, partial2 = 1770}
+```
+
+In the type of `addTotal`, the part `{total: Num | a}` expresses exactly what we
+wanted: the argument must have a field `total: Num`, but the *tail* (the rest of
+the record type) is polymorphic, and `a` may be substituted for arbitrary fields
+(such as `jan: Num, feb: Num`). We used two different generic parameters `a` and
+`b`, to express that the tails of the arguments may differ.  If we used `a` in
+both places, as in `forall a. {total: Num | a} -> {total: Num | a} -> Num`, we
+could still write `addTotal {total = 1, foo = 1} {total = 2, foo = 2}` but not
+`addTotal {total = 1, foo = 1} {total = 2, bar = 2}`. Using distinct parameters
+`a` and `b` gives us maximum flexibility.
+
+What comes before the tail may include several fields, is in e.g. `forall a.
+{total: Num, subtotal: Num | a} -> Num`.
```

Comment:
There lacks an example where the row variable is actually used.

Something like
```
f : forall r. { a : Num, b : Num | r } -> {  a : Num, b : Num, sum-a-b : Num | r }
f x = x & { sum-a-b  = x.a + x.b }
```

Reply:
It's a good example, but sadly merging isn't typable currently, so this won't typecheck. For now I'll use the extension operator. We'll revisit the manual anyway just before release, if only because syntax is likely to have changed in between.